# T Test

In [1]:
%reload_ext autoreload
%autoreload 2

In [29]:
from pkg_resources import importlib

# from ohdsi import common
from ohdsi.database_connector import connect
from ohdsi import database_connector
from ohdsi import common

# import warnings
# warnings.simplefilter(action='ignore', category=UserWarning)

import pandas as pd
import matplotlib.pyplot as plt
pd.options.mode.chained_assignment = None  # default='warn'

## OMOP database
### Get connection details

### Connect to the database

In [30]:
conn = connect(
    dbms = "postgresql",
    connection_string = "jdbc:postgresql://localhost:5432/omopdb",
    user = "ohdsi",
    password = "ohdsi",
)

Connecting using PostgreSQL driver


## Extract the cohort data

In [37]:
sessions = importlib.import_module("v6-sessions")

In [38]:
# rps_cohort
sql = "SELECT person_id FROM omopcdm.measurement WHERE measurement_concept_id in (36770706)"
df = database_connector.query_sql(conn, sql)
df = common.convert_from_r(df)
print(df)
pts_RPS = list(df.person_id.values)
print(pts_RPS)

rps_cohort_original = sessions.cohort.__create_cohort_dataframe(
    connection=conn,
    patient_ids=pts_RPS,
    features="sarcoma"
)
rps_cohort = rps_cohort_original.to_pandas()

# pelvis_cohort
sql = "SELECT person_id FROM omopcdm.measurement WHERE measurement_concept_id in (36768752)"
df = database_connector.query_sql(conn, sql)
df = common.convert_from_r(df)
pts_Pelvis = list(df.person_id.values)

pelvis_cohort_original = sessions.cohort.__create_cohort_dataframe(
    connection=conn,
    patient_ids=pts_Pelvis,
    features="sarcoma"
)
pelvis_cohort = pelvis_cohort_original.to_pandas()

# rps_pelvis_cohort
sql = "SELECT person_id FROM omopcdm.measurement WHERE measurement_concept_id in (36770706, 36768752)"
df = database_connector.query_sql(conn, sql)
df = common.convert_from_r(df)
pts_RPS_Pelvis = list(df.person_id.values)

rps_pelvis_cohort_original = sessions.cohort.__create_cohort_dataframe(
    connection=conn,
    patient_ids=pts_RPS_Pelvis,
    features="sarcoma"
)
rps_pelvis_cohort = rps_pelvis_cohort_original.to_pandas()


print(f'RPS cohort: {len(rps_cohort)}')
print(f'Pelvis cohort: {len(pelvis_cohort)}')
print(f'RPS + Pelvis cohort: {len(rps_pelvis_cohort)}')

     person_id
1          1.0
2          3.0
3          4.0
4          5.0
5          6.0
..         ...
302      591.0
303      592.0
304      595.0
305      597.0
306      598.0

[306 rows x 1 columns]
[1.0, 3.0, 4.0, 5.0, 6.0, 9.0, 12.0, 14.0, 18.0, 20.0, 21.0, 24.0, 27.0, 29.0, 30.0, 32.0, 38.0, 39.0, 41.0, 42.0, 43.0, 45.0, 46.0, 47.0, 49.0, 50.0, 53.0, 54.0, 55.0, 56.0, 57.0, 61.0, 62.0, 63.0, 64.0, 65.0, 66.0, 73.0, 74.0, 76.0, 77.0, 80.0, 84.0, 85.0, 86.0, 87.0, 88.0, 92.0, 93.0, 94.0, 97.0, 98.0, 99.0, 102.0, 105.0, 106.0, 112.0, 116.0, 117.0, 120.0, 121.0, 122.0, 125.0, 126.0, 127.0, 129.0, 131.0, 132.0, 133.0, 137.0, 138.0, 141.0, 143.0, 144.0, 146.0, 151.0, 152.0, 153.0, 154.0, 155.0, 156.0, 157.0, 160.0, 161.0, 163.0, 167.0, 168.0, 169.0, 170.0, 171.0, 173.0, 174.0, 175.0, 176.0, 178.0, 180.0, 181.0, 182.0, 184.0, 188.0, 190.0, 192.0, 194.0, 202.0, 203.0, 207.0, 210.0, 211.0, 212.0, 219.0, 222.0, 223.0, 224.0, 225.0, 226.0, 228.0, 229.0, 230.0, 231.0, 235.0, 237.0, 238.0, 

/Users/frankmartin/Repositories/idea4rc-vantage6-algorithms/v6-sessions/v6-sessions/cohort.py:113: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  converted_df = converted_df.applymap(


info > Dropped 0 rows
info > Converting column types
info > -->  Done
info > Loading SQL file: sarcoma
info > -->  Done
info > Injecting patient IDs into SQL
info > -->  Done
info > Executing SQL
info > Converting dataframe to pandas


/Users/frankmartin/Repositories/idea4rc-vantage6-algorithms/v6-sessions/v6-sessions/cohort.py:113: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  converted_df = converted_df.applymap(


info > Dropped 0 rows
info > Converting column types
info > -->  Done
info > Loading SQL file: sarcoma
info > -->  Done
info > Injecting patient IDs into SQL
info > -->  Done
info > Executing SQL
info > Converting dataframe to pandas


/Users/frankmartin/Repositories/idea4rc-vantage6-algorithms/v6-sessions/v6-sessions/cohort.py:113: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  converted_df = converted_df.applymap(


info > Dropped 0 rows
info > Converting column types
info > -->  Done
RPS cohort: 306
Pelvis cohort: 294
RPS + Pelvis cohort: 600


In [39]:
rps_cohort.columns

Index(['patient_id', 'age', 'sex', 'censor', 'status', 'survival_days',
       'histology', 'fnclcc_grade', 'tumor_size', 'surgery_date',
       'surgery_concept', 'multifocality', 'completeness_of_resection',
       'completeness_of_resection_concept_id', 'tumor_rupture',
       'pre_operative_chemo', 'post_operative_chemo', 'pre_operative_radio',
       'post_operative_radio', 'local_recurrence', 'distant_metastasis',
       'n_cancer_episodes'],
      dtype='object')

In [40]:
rps_cohort.dtypes

patient_id                                          float64
age                                                 float64
sex                                                category
censor                                                 bool
status                                             category
survival_days                                       float64
histology                                          category
fnclcc_grade                                       category
tumor_size                                          float64
surgery_date                            datetime64[ns, UTC]
surgery_concept                                     float64
multifocality                                      category
completeness_of_resection                          category
completeness_of_resection_concept_id                float64
tumor_rupture                                          bool
pre_operative_chemo                                    bool
post_operative_chemo                    

## MockClient
We mock to have two organizations with three databases each (<code>rps_cohort</code>, <code>pelvis_cohort</code> and <code>rps_pelvis_cohort</code>). <br><br>
The first organization has:
* <code>rps_cohort[:10]</code> (10 pts)
* <code>pelvis_cohort[:10]</code> (10 pts)
* <code>rps_pelvis_cohort[:10]</code> (10 pts)

The second organization has:
* <code>rps_cohort[10:]</code> (304 pts)
* <code>pelvis_cohort[10:]</code> (276 pts)
* <code>non_liposarcoma_cohort[10:]</code> (590 pts)

In [41]:
from vantage6.algorithm.mock.network import MockNetwork

network = MockNetwork(
    "v6-analytics",
    datasets=[
        {
            "rps": {"database": rps_cohort[:10], "db_type": "omop"},
            "pelvis": {"database": pelvis_cohort[:10], "db_type": "omop"},
            "rps_pelvis": {"database": rps_pelvis_cohort[:10], "db_type": "omop"}
        },
        {
            "rps": {"database": rps_cohort[10:], "db_type": "omop"},
            "pelvis": {"database": pelvis_cohort[10:], "db_type": "omop"},
            "rps_pelvis": {"database": rps_pelvis_cohort[10:], "db_type": "omop"}
        }
    ],
    collaboration_id=1,
)

client = network.user_client

info > Using dataframes decorator
info > Using dataframes decorator
info > Using dataframes decorator
info > Using dataframes decorator
info > Using dataframes decorator
[info ] - Mock network created with 2 nodes


In [42]:
# overall summary results
task = client.task.create(
    method="t_test_central",
    organizations=[1],
    arguments={
        "organizations_to_include": [1,2]
    },
    databases=[
        [
            {"type": "dataframe", "dataframe_id": 1},
            {"type": "dataframe", "dataframe_id": 2},
            {"type": "dataframe", "dataframe_id": 3}
        ]
    ],
    action="central_compute"
)

[info ] - Validating function action: central_compute
info > Creating subtask for all organizations in the collaboration
[info ] - Validating function action: federated_compute
info > Checking number of records in the DataFrame.
info > Computing mean for patient_id
info > Computing sample variance for patient_id
info > Computing mean for age
info > Computing sample variance for age
info > Computing mean for survival_days
info > Skipping survival_days due to insufficient data.
info > Computing mean for tumor_size
info > Computing sample variance for tumor_size
info > Computing mean for surgery_concept
info > Computing sample variance for surgery_concept
info > Computing mean for completeness_of_resection_concept_id
info > Computing sample variance for completeness_of_resection_concept_id
info > Computing mean for n_cancer_episodes
info > Computing sample variance for n_cancer_episodes
info > Checking number of records in the DataFrame.
info > Computing mean for patient_id
info > Computi

info > Computing mean for n_cancer_episodes
info > Computing sample variance for n_cancer_episodes
[info ] - Validating function action: federated_compute
info > Checking number of records in the DataFrame.
info > Computing mean for patient_id
info > Computing sample variance for patient_id
info > Computing mean for age
info > Computing sample variance for age
info > Computing mean for survival_days
info > Skipping survival_days due to insufficient data.
info > Computing mean for tumor_size
info > Computing sample variance for tumor_size
info > Computing mean for surgery_concept
info > Computing sample variance for surgery_concept
info > Computing mean for completeness_of_resection_concept_id
info > Computing sample variance for completeness_of_resection_concept_id
info > Computing mean for n_cancer_episodes
info > Computing sample variance for n_cancer_episodes
info > Checking number of records in the DataFrame.
info > Computing mean for patient_id
info > Computing sample variance for

In [43]:
results = client.result.from_task(task_id=task.get("id"))
results

[{'rps': {'patient_id': {'p_value': 4.233732697755954e-08,
    't_score': -5.623852902476283},
   'age': {'p_value': 0.13679522594033244, 't_score': -1.491774531093365},
   'tumor_size': {'p_value': 0.6293395843983021,
    't_score': -0.4832020293582413},
   'completeness_of_resection_concept_id': {'p_value': 0.14618713724033738,
    't_score': -1.4571516317417652}},
  'pelvis': {'patient_id': {'p_value': 3.368435552708604e-08,
    't_score': -5.6734648761692394},
   'age': {'p_value': 0.24716726174728043, 't_score': 1.1595797362803264},
   'tumor_size': {'p_value': 0.431701737086863,
    't_score': -0.7874896128939661},
   'completeness_of_resection_concept_id': {'p_value': 0.07731201730832393,
    't_score': -1.7735812807020452}},
  'rps_pelvis': {'patient_id': {'p_value': 4.0527341482743395e-08,
    't_score': -5.560763309418322},
   'age': {'p_value': 0.2676265967188005, 't_score': -1.1095787341674967},
   'tumor_size': {'p_value': 0.6459394456979988,
    't_score': -0.459668616050

In [47]:
for item in results:
    print(pd.DataFrame(item))


                                                                                    rps  \
patient_id                            {'p_value': 4.233732697755954e-08, 't_score': ...   
age                                   {'p_value': 0.13679522594033244, 't_score': -1...   
tumor_size                            {'p_value': 0.6293395843983021, 't_score': -0....   
completeness_of_resection_concept_id  {'p_value': 0.14618713724033738, 't_score': -1...   

                                                                                 pelvis  \
patient_id                            {'p_value': 3.368435552708604e-08, 't_score': ...   
age                                   {'p_value': 0.24716726174728043, 't_score': 1....   
tumor_size                            {'p_value': 0.431701737086863, 't_score': -0.7...   
completeness_of_resection_concept_id  {'p_value': 0.07731201730832393, 't_score': -1...   

                                                                             rps_pelvis 